### PR TITLE
Fix pymc sampler

### LINKED ISSUE #928: pymc sampler appears to not sample correctly
In the sampler study notebook, the 1-dim bimodal seems to not be sampled correctly (the pymc3 version worked) https://github.com/ICB-DCM/pyPESTO/blob/develop/doc/example/sampler_study.ipynb. Current develop branch, not merged to main yet.

image

### PR DESCRIPTION
**Pull request by yannikschaelte**
Use different construction of aesara operator, fixes #928

Also: Fix deprecation warning (scipy optimize.minimize maxiter -> maxfun)

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -108,7 +108,7 @@ description =
 
 [testenv:notebooks2]
 allowlist_externals = bash
-extras = example,amici,petab,pymc3,pymc,emcee,select
+extras = example,amici,petab,pymc,pymc,emcee,select
```
Comment (FFroehlich) [inline tox.ini:111]:
```suggestion
extras = example,amici,petab,pymc,emcee,select
```

File: pypesto/sample/pymc.py [hunk 2]
```diff
@@ -139,8 +178,8 @@ def sample(self, n_samples: int, coeff: float = 1.0):
 
         # create model context
         with pymc.Model():
-            # uniform bounds
-            k = [
+            # uniform prior
```
Comment (FFroehlich) [inline pypesto/sample/pymc.py:181]:
```suggestion
            # parameter bounds as uniform prior
```

File: pypesto/sample/pymc.py [hunk 3]
```diff
@@ -2,56 +2,93 @@
...
-            "python packages aesara, arviz and pymc. Please install "
-            "these packages via `pip install aesara arviz pymc`."
-        )
+# implementation based on:
+# https://www.pymc.io/projects/examples/en/latest/case_studies/blackbox_external_likelihood_numpy.html
 
 
-class AesaraDist(pymc.NoDistribution):
-    """PyMC distribution wrapper for AesaraObjectiveRVs."""
+class AesaraObjectiveOp(at.Op):
```
Comment (FFroehlich) [inline pypesto/sample/pymc.py:24]: seems pretty redundant with https://github.com/ICB-DCM/pyPESTO/blob/488eca74bf5b6abe0a0c60de0215997c6973c0a8/pypesto/objective/aesara.py#L151

File: pypesto/sample/pymc.py [hunk 4]
```diff
@@ -2,56 +2,93 @@
...
+
+    def grad(self, inputs, g):  # noqa
+        """Calculate the vector-Jacobian product."""
+        # the method that calculates the gradients - it actually returns the
+        # vector-Jacobian product - g[0] is a vector of parameter values
+        (theta,) = inputs  # our parameters
+        return [g[0] * self._log_prob_grad(theta)]
+
+
+class AesaraGradientOp(at.Op):
```
Comment (FFroehlich) [inline pypesto/sample/pymc.py:76]: see https://github.com/ICB-DCM/pyPESTO/blob/488eca74bf5b6abe0a0c60de0215997c6973c0a8/pypesto/objective/aesara.py#L195

**Review by FFroehlich — APPROVED** [review]

File: pypesto/sample/pymc.py (hunk 3 above)
Comment (yannikschaelte) [reply pypesto/sample/pymc.py:24]: it certainly is. however, I could not get the pypesto/objective approach to run, kept running into errors, partly induced by recent changes in pymc. this one here is now based on a standard template for blackbox objectives (referenced in the code). :grimacing:

Comment (FFroehlich) [reply pypesto/sample/pymc.py:24]: Ah, okay. Then let me have another look tomorrow. Would prefer not to duplicate functionality and have the caching implemented in AmiciObjective (which might be the actual problem).

**Comment by FFroehlich** [comment]
Doesn't look like the aesara objective methods can easily be reused, so just did some additional cleanup & renaming.

**Comment by yannikschaelte** [comment]
Thanks! Unifying both should be possible, but then letting it be until esp. the pymc docs have been further updated from pymc3.

**Pull request merged**